Fixing join tests with analyzer

vdimir opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)


TODO:
- [x] https://github.com/ClickHouse/ClickHouse/pull/49359


Fixed reference files:

* `01353_low_cardinality_join_types`

	```sql
	-- Without Analyzer LowCardinality(String)
	-- With Analyzer: String
	SELECT toTypeName(s) FROM (SELECT toLowCardinality('a') as s) t1 JOIN (SELECT 'a' as s) t2 USING (s);
	
	-- Without Analyzer LowCardinality(String)
	-- With Analyzer: String
    SELECT toTypeName(t1.s) FROM (SELECT toLowCardinality('a') as s) t1 JOIN (SELECT 'a' as s) t2 USING (s);

    -- Without Analyzer String
    -- With Analyzer: String
    SELECT toTypeName(t2.s) FROM (SELECT toLowCardinality('a') as s) t1 JOIN (SELECT 'a' as s) t2 USING (s);
    ```

* `02267_join_dup_columns_issue36199`

    ```sql
    -- Without Analyzer return only columns `count(), x` from right table
    -- With Analyzer return only columns `count('x'), count('y'), x` from right table
    SELECT * FROM ( SELECT 2 AS x ) AS t1 RIGHT JOIN ( SELECT count('x'), count('y'), 2 AS x ) AS t2 ON t1.x = t2.x;
    ```

* `02242_join_rocksdb`

    Fixed bug and adjusted error code in test

* `01049_join_low_card_bug_long`

	Fix bug with query:

    ```sql
    SELECT t.t from (SELECT 1 as t) as a FULL JOIN (SELECT 1 as t) as t USING t;
    ```

    adjust reference file (columns after USING have type of supertype of the source columns), also it needs fix from https://github.com/ClickHouse/ClickHouse/pull/49359 (so, it's not removed from broken_tests.txt)

* `01721_join_implicit_cast_long`
    Adjust reference.
	
	Not sure about the change:
	
	```sql
	-- Before: Nullable
	-- With analyzer: not-Nullable
	SELECT toTypeName(a) FROM (SELECT 1 as a ) t1 RIGHT JOIN  (SELECT 1 as a) t2 USING (a) SETTINGS join_use_nulls = 1;
	```

* `01062_pm_all_join_with_block_continuation`
        ~~Fixed test file, JOIN using alias from external scope.~~ Removed this fix from the PR, will in another one.

* `02000_join_on_const`
        Fixed test file, `... ANY JOIN ... ON 1 == 1` is forbidden

* `02382_join_and_filtering_set`
    Copy missing corresponding code from Interpreter, adjust test (removed with cube)

* `01890_materialized_distributed_join`
    Function getHeaderForProcessingStage handle join on top of query tree

vdimir committed:
Fix 02267_join_dup_columns_issue36199

vdimir committed:
Fix 01353_low_cardinality_join_types

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #49555 from ClickHouse/ClickHouse:vdimir/analyzer_fix_join_tests -->
This is an automated comment for commit a291d7c8315d5c7bdf78e1f49b0b92f7940184f9 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49555/a291d7c8315d5c7bdf78e1f49b0b92f7940184f9/ci_running.html)
The overall status of the commit is 🟢 success

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🟢 success</td></tr>
</table>

vdimir committed:
Fix 02242_join_rocksdb

vdimir committed:
[wip] Fixing 01049_join_low_card_bug_long

vdimir committed:
Fix 01049_join_low_card_bug_long

vdimir committed:
Fix 01721_join_implicit_cast_long

vdimir commented:
Some test broken on old analyzer

vdimir committed:
Fix 01062_pm_all_join_with_block_continuation

Fixed test file, JOIN using alias from external scope

vdimir committed:
Fix 02000_join_on_const

vdimir commented:
- [ ] https://s3.amazonaws.com/clickhouse-test-reports/49555/1e50db004d10d0322494635f60897e7c1a963c84/fuzzer_astfuzzerasan/report.html

vdimir committed:
Fix 02382_join_and_filtering_set

Copy missing corresponding code from Interpreter, adjust test (removed
with cube)

vdimir committed:
Fix 01890_materialized_distributed_join

Function getHeaderForProcessingStage handle join on top of query tree

vdimir committed:
fix style check

vdimir commented:
> Stateless tests flaky check (asan) — fail: 21, passed: 193 [Details](https://s3.amazonaws.com/clickhouse-test-reports/49555/4af8d68d2aad3553137945c5c632dd308070db2a/stateless_tests_flaky_check__asan_.html)
> 01062_pm_all_join_with_block_continuation:                              [ FAIL ] 147.69 sec. - Test runs too long (> 60s). Make it faster

Normally it takes <50sec: 
https://play.clickhouse.com/play?user=play#U0VMRUNUCiAgICBxdWFudGlsZXMoMC41LCAwLjksIDAuOTUpKHRlc3RfZHVyYXRpb25fbXMgLyAxMDAwKSBGUk9NIGNoZWNrcwpXSEVSRSB0ZXN0X25hbWUgPSAnMDEwNjJfcG1fYWxsX2pvaW5fd2l0aF9ibG9ja19jb250aW51YXRpb24nCkFORCBjaGVja19uYW1lIGxpa2UgJyVhc2FuJScKQU5EIGNoZWNrX3N0YXJ0X3RpbWUgPD0gbm93KCkgLSBJTlRFUlZBTCAxMCBEQVkKQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwCgo=

vdimir committed:
Revert "Fix 01062_pm_all_join_with_block_continuation"

This reverts commit 88865ace8efb3c10307ddce088a7c44ebb7009fd.

vdimir commented:
> Stateless tests flaky check (asan) — fail: 1, passed: 260 [Details](https://s3.amazonaws.com/clickhouse-test-reports/49555/964c7f95228eab5117119e190825f660e503f589/stateless_tests_flaky_check__asan_.html)

> 01721_join_implicit_cast_long:                                          [ FAIL ] 601.03 sec. - Timeout!

It doesn't have `long` tag in the code, I'm not sure if we still count test as long by filename.
It became twice longer that it is (added run it with and without analyzer). Another option is just to split it into two tests.

All in all, this PR is finished it contains fixes for several tests. What was done for each test is described in the  top-level PR comment, and also each one is addressed in separate commit.

@novikd you may take a look

alexey-milovidov commented:
The new test appears to be longer than 600 seconds under some settings.

novikd reviewed (commented):
In general LGTM

vdimir committed:
fix

vdimir commented on tests/queries/0_stateless/01049_join_low_card_bug_long.sql.j2:
```diff
@@ -98,13 +100,13 @@ SET join_use_nulls = 0;
 
 SELECT lc, toTypeName(lc)  FROM l_lc AS l RIGHT JOIN r_lc AS r USING (x) ORDER BY l.lc;
 
-SELECT lowCardinalityKeys(lc.lc) FROM r FULL JOIN l_lc as lc USING (lc) ORDER BY lowCardinalityKeys(lc.lc);
```
In new analyzer `lc.lc` became String (because r contains `String` and we take supertype as a result type), so lowCardinalityKeys is not applicable. We can leave this query only for old analyzer, but I don't think it's essential.

vdimir commented on src/Interpreters/getHeaderForProcessingStage.cpp:
```diff
@@ -124,13 +126,27 @@ Block getHeaderForProcessingStage(
-                removeJoin(*query->as<ASTSelectQuery>(), new_rewriter_result, context);
+                {
+                    if (!query_info.planner_context)
+                        throw Exception(ErrorCodes::LOGICAL_ERROR, "Query is not analyzed");
```
Updated to `Query is not analyzed: no planner context`. Either way, it's `LOGICAL_ERROR` and it doesn't matter a lot for a user, but rather for developer.

vdimir commented:
> Stress test (tsan) — check_status.tsv doesn't exists [Details](https://s3.amazonaws.com/clickhouse-test-reports/49555/4cfe6002f82699c83e72776bdc3a3f89f005b0bc/stress_test__tsan_.html)
> Upgrade check (tsan) — check_status.tsv doesn't exists [Details](https://s3.amazonaws.com/clickhouse-test-reports/49555/4cfe6002f82699c83e72776bdc3a3f89f005b0bc/upgrade_check__tsan_.html)

https://github.com/ClickHouse/ClickHouse/issues/49076

> Stateless tests (release, analyzer) — fail: 1, passed: 5649, skipped: 48 [Details](https://s3.amazonaws.com/clickhouse-test-reports/49555/4cfe6002f82699c83e72776bdc3a3f89f005b0bc/stateless_tests__release__analyzer_.html)

`02886_missed_json_subcolumns` probably related, will check

vdimir commented:
> Integration tests (release) [3/4] — fail: 1, passed: 603 [Details](https://s3.amazonaws.com/clickhouse-test-reports/49555/3aac663bd2655360d503d4213fddbde3beaeb1df/integration_tests__release__[3_4].html)

Seems `test_replicated_zero_copy_projection_mutation/test.py::test_hardlinks_preserved_when_projection_dropped` is a bit flaky.
No other failures ([ci db](https://play.clickhouse.com/play?user=play#CgoKU0VMRUNUCiAgICBzdWJzdHIoY29tbWl0X3NoYSwgMSw5KSBhcyBzaGEsCiAgICBjaGVja19uYW1lLAogICAgdGVzdF9uYW1lLAogICAgZm9ybWF0UmVhZGFibGVUaW1lRGVsdGEoaW50RGl2KG5vdygpIC0gY2hlY2tfc3RhcnRfdGltZSwgMzYwMCkgKiAzNjAwKSBhcyBhZ28sCiAgICByZXBvcnRfdXJsCkZST00gY2hlY2tzCldIRVJFIDEKICAgIEFORCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgNyBEQVkKCiAgICBBTkQgdGVzdF9zdGF0dXMgIT0gJ1NLSVBQRUQnCiAgICBBTkQgdGVzdF9zdGF0dXMgTElLRSAnRiUnCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9yZXBsaWNhdGVkX3plcm9fY29weV9wcm9qZWN0aW9uX211dGF0aW9uJScKT1JERVIgQlkgIGNoZWNrX3N0YXJ0X3RpbWUgREVTQw==)), but issue related to timeouts in tests



<details>
<summary>Details</summary>

We deleted objects after timeout:

```
2023-10-12 12:23:59 [ 601 ] DEBUG : http://172.16.11.8:9001 "GET /root?delimiter=&encoding-type=url&list-type=2&max-keys=1000&prefix=data%2F HTTP/1.1" 200 0 (connectionpool.py:546, _make_request)
2023-10-12 12:23:59 [ 601 ] INFO : list_objects (20): ['data/bvw/wrljevlduzmgtpddbnjdicmjbqmul', 'data/gar/bdzxvruhcbojwnssexrlvsuoknmyg', 'data/ixn/uqsdxqevkpebrpidxuywczmsmwhlb', 'data/kol/jzwsrnushtokftdzaloolthtxldvp', 'data/kvs/uhzthybwkhsxuiteqoyvdlniitnfb', 'data/lhi/ctdysdzbymnllpxcqffbyrnxjonwf', 'data/loi/nkwrplbttcgzamiwqmedqbfgqcxvk', 'data/lsx/iyuqumlazclzrdsitaobgcbdzookw', 'data/mbx/gqvwlivusmkrehndgnbnjkochywdc', 'data/mqk/dljgyadfnccqijfgudkkgnkssrnej', 'data/nzb/fhfkriolhlgrjcskicmcyivmfjegy', 'data/ood/deqohcrddwnsademgifqbmvunctgq', 'data/pnd/kbdarptmoapjfglrklaeevnaqtkex', 'data/rey/tesbqzzygjsnwrvvrbvesjlscbetv', 'data/seq/wlsptkdxjzjxwnvesxhuyltndvbcp', 'data/sya/uaggjvudvwbtgvbvqmromayzzupdc', 'data/vie/vtakpllynnxwkykfmsghonyqpeqra', 'data/vla/bsrahqodprsluqzngqkuivtmrcebm', 'data/vtd/vehexvssfhqseohjruybicspupwcl', 'data/xts/jusayraucumkmwegngwsgdyenokeu'] (test.py:75, list_objects)
2023-10-12 12:24:00 [ 601 ] DEBUG : http://172.16.11.8:9001 "GET /root?delimiter=&encoding-type=url&list-type=2&max-keys=1000&prefix=data%2F HTTP/1.1" 200 0 (connectionpool.py:546, _make_request)
2023-10-12 12:24:00 [ 601 ] INFO : list_objects (20): ['data/bvw/wrljevlduzmgtpddbnjdicmjbqmul', 'data/gar/bdzxvruhcbojwnssexrlvsuoknmyg', 'data/ixn/uqsdxqevkpebrpidxuywczmsmwhlb', 'data/kol/jzwsrnushtokftdzaloolthtxldvp', 'data/kvs/uhzthybwkhsxuiteqoyvdlniitnfb', 'data/lhi/ctdysdzbymnllpxcqffbyrnxjonwf', 'data/loi/nkwrplbttcgzamiwqmedqbfgqcxvk', 'data/lsx/iyuqumlazclzrdsitaobgcbdzookw', 'data/mbx/gqvwlivusmkrehndgnbnjkochywdc', 'data/mqk/dljgyadfnccqijfgudkkgnkssrnej', 'data/nzb/fhfkriolhlgrjcskicmcyivmfjegy', 'data/ood/deqohcrddwnsademgifqbmvunctgq', 'data/pnd/kbdarptmoapjfglrklaeevnaqtkex', 'data/rey/tesbqzzygjsnwrvvrbvesjlscbetv', 'data/seq/wlsptkdxjzjxwnvesxhuyltndvbcp', 'data/sya/uaggjvudvwbtgvbvqmromayzzupdc', 'data/vie/vtakpllynnxwkykfmsghonyqpeqra', 'data/vla/bsrahqodprsluqzngqkuivtmrcebm', 'data/vtd/vehexvssfhqseohjruybicspupwcl', 'data/xts/jusayraucumkmwegngwsgdyenokeu'] (test.py:75, list_objects)
2023-10-12 12:24:00 [ 601 ] DEBUG : Executing query DROP TABLE IF EXISTS test_hardlinks_preserved_when_projection_dropped SYNC on node1 (cluster.py:3409, query)
2023-10-12 12:24:00 [ 601 ] DEBUG : Executing query DROP TABLE IF EXISTS test_hardlinks_preserved_when_projection_dropped SYNC on node2 (cluster.py:3409, query)
```

```
2023.10.12 12:24:00.270615 [ 371 ] {} <Debug> S3ObjectStorage: Objects with paths [data/vie/vtakpllynnxwkykfmsghonyqpeqra, data/nzb/fhfkriolhlgrjcskicmcyivmfjegy, data/rey/tesbqzzygjsnwrvvrbvesjlscbetv, data/lhi/ctdysdzbymnllpxcqffbyrnxjonwf, data/bvw/wrljevlduzmgtpddbnjdicmjbqmul, data/sya/uaggjvudvwbtgvbvqmromayzzupdc, data/seq/wlsptkdxjzjxwnvesxhuyltndvbcp, data/vtd/vehexvssfhqseohjruybicspupwcl, data/mqk/dljgyadfnccqijfgudkkgnkssrnej] were removed from S3
```


And expected objects:


```
list_objects (2): ['data/gar/bdzxvruhcbojwnssexrlvsuoknmyg', 'data/mbx/gqvwlivusmkrehndgnbnjkochywdc'] (test.py:75, list_objects)
```

Was deleted only after test was finished and table was dropped.

```
node1/logs/clickhouse-server.log
517:2023.10.12 12:22:53.716436 [ 12 ] {ea2e0500-08fa-4c29-95f7-4297baca346a} <Debug> executeQuery: (from 172.16.11.1:55232) DROP TABLE IF EXISTS test_hardlinks_preserved_when_projection_dropped SYNC (stage: Complete)
2037:2023.10.12 12:24:00.196979 [ 12 ] {8db3cd4d-2ed7-4a21-9e47-2ae7feff10df} <Debug> executeQuery: (from 172.16.11.1:55710) DROP TABLE IF EXISTS test_hardlinks_preserved_when_projection_dropped SYNC (stage: Complete)

node2/logs/clickhouse-server.log
907:2023.10.12 12:24:00.260354 [ 361 ] {205de67e-686c-4b76-8d04-21e08a19b434} <Debug> executeQuery: (from 172.16.11.1:64916) DROP TABLE IF EXISTS test_hardlinks_preserved_when_projection_dropped SYNC (stage: Complete)
...

2023.10.12 12:24:00.201130 [ 163 ] {} <Debug> S3ObjectStorage: Objects with paths [data/mbx/gqvwlivusmkrehndgnbnjkochywdc] were removed from S3
...
2023.10.12 12:24:00.275343 [ 371 ] {} <Debug> S3ObjectStorage: Objects with paths [data/gar/bdzxvruhcbojwnssexrlvsuoknmyg] were removed from S3

```

So I believe no bug here


</details>
